refactor: modular controller init

matthewwalsh0 opened the pull request:
## **Description**

Architecture to support modular controller initialisation to avoid the current monolithic `metamask-controller.js`.

Includes:

- `PPOMController` as a simple example.
- `TransactionController` as perhaps the most complex example.

### Overview

Uses isolated functions (in `app/scripts/controller-init/`) with a generic `ControllerInitFunction` type, each of which returns the following given a standardised and generic `ControllerInitRequest`:

- Required `controller` instance.
- Optional `api` object to declare background API functions.
- Optional `persistedStateKey` to specify persisted state key, or disable.
- Optional `memStateKey` to specify mem-store state key, or disable.

Dependent controllers can be retrieved via a type safe `getController` function in the `ControllerInitRequest` to provide a standard error message if controller is not yet initialised.

Messengers are intentionally defined and assigned in separate isolated files to allow alternate code ownership, and not over-power the controller logic and introduce bugs.

Initialisation logic is encapsulated in `app/scripts/controller-init/utils.ts`, which iterates over the functions to construct the controller instances, plus generate the API and store properties.

### Advantages

- Backwards compatible with almost no changes to legacy controller initialisation.
- Facilitates an iterative approach meaning teams can refactor individual controllers ad-hoc.
- Facilitates iterative Typescript adoption of controller initialisation code.
- Supports modular unit testing of controller / messenger event handlers.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28948?quickstart=1)

## **Related issues**

## **Manual testing steps**

## **Screenshots/Recordings**

### **Before**

### **After**

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

matthewwalsh0 committed:
Initial draft

matthewwalsh0 committed:
Support modular API functions

matthewwalsh0 committed:
Merge branch 'main' into refactor/modular-controller-init

matthewwalsh0 committed:
Add state key methods

matthewwalsh0 committed:
Improve types

matthewwalsh0 committed:
Rename file

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -7474,4 +7517,82 @@ export default class MetamaskController extends EventEmitter {
     return this.networkController.state.selectedNetworkClientId;
   }
+
+  #initControllers({ initObjects, initState }) {
```
Both these functions could go in a separate `app/scripts/controller-init/utils.ts` for better encapsulation and testing, and TypeScript.

matthewwalsh0 committed:
Add PPOM controller init class

matthewwalsh0 committed:
Update ownership

matthewwalsh0 committed:
Fix E2E tests

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -7474,4 +7485,83 @@ export default class MetamaskController extends EventEmitter {
     return this.networkController.state.selectedNetworkClientId;
   }
+
+  #initControllers({ initObjects, initState }) {
```
These two functions could go in a `app/scripts/controller-init/utils.ts` and we pass the init request as an argument.

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -478,1922 +451,2104 @@ export default class MetamaskController extends EventEmitter {
-      'user-operation-added',
-      this._onUserOperationAdded.bind(this),
-    );
+    const controllerInitObjects = [
```
In future, we could automatically order this list via each init class explicitly declaring it's dependencies with a new abstract method such as `getControllerDependencies`.

Alternatively, the order may not matter if we removed all hard references between the controllers, and relied solely on the controller messenger.

matthewwalsh0 committed:
Move messengers to separate files

matthewwalsh0 committed:
Add flat state

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Minimise diff

matthewwalsh0 committed:
Remove duplicate code

matthewwalsh0 committed:
Merge branch 'main' into refactor/modular-controller-init

matthewwalsh0 committed:
Split messenger types

matthewwalsh0 committed:
Make messengers optional

Add legacy init class.
Add JSDoc.

matthewwalsh0 committed:
Fix E2E tests

matthewwalsh0 committed:
Add initial unit tests

matthewwalsh0 committed:
Add utils unit tests

matthewwalsh0 committed:
Add unit tests

matthewwalsh0 committed:
Merge branch 'main' into refactor/modular-controller-init

dbrans commented on app/scripts/controller-init/messengers/transaction-controller-messenger.ts:
```diff
@@ -0,0 +1,92 @@
+import { AccountsControllerGetSelectedAccountAction } from '@metamask/accounts-controller';
```
Hey Matt, I noticed that the messengers and controllers are in separate folders, even though they seem tightly coupled. Curious about the reasoning behind that.

matthewwalsh0 replied:
From the PR description:

>Messengers are intentionally defined in separate isolated files to allow alternate code ownership, and not over-power the controller logic and introduce bugs.

Traditionally, the messengers are defined inline with the controllers but that means any controller can casually connect or start depending on another, without any oversight or awareness from other teams.

Meaning it's a common source of bugs, and so defining them separately ensures the controller initialisation doesn't have "sudo" permissions to the entire system, but we can instead define their "permissions" elsewhere and tell each controller and initialisation instance what it's allowed to do.

Practically, it also means the platform team for example could own the messenger directories to stay informed of the high level connections between controllers.

dbrans replied:
Thank you for the extra context!

dbrans commented:
Hey Matt, love this idea—breaking up metamask-controller.js and adding types to the controller initialization would be a huge win. Thanks for suggesting it!

matthewwalsh0 committed:
Merge branch 'main' into refactor/modular-controller-init

matthewwalsh0 committed:
Merge branch 'main' into refactor/modular-controller-init

HowardBraham commented:
@davidmurdoch I just did a circular dependencies analysis on this PR.

These circular deps were removed:
- app/scripts/controllers/swaps/swaps.constants.ts > app/scripts/controllers/swaps/swaps.types.ts

These circular deps were added:
- app/scripts/controllers/swaps/swaps.types.ts > app/scripts/controllers/swaps/index.ts
- app/scripts/controllers/swaps/swaps.types.ts > app/scripts/controllers/swaps/index.ts > app/scripts/controllers/swaps/swaps.utils.ts

Though I'm a little confused how this happened at all, when none of these files were changed 🤷

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -5704,6 +5562,7 @@ export default class MetamaskController extends EventEmitter {
         handleUpdate();
       },
       getStatePatches: () => patchStore.flushPendingPatches(),
+      ...this.controllerApi,
```
Nit: Perhaps we could move this to before the patch methods, so they can't be accidentally overwritten? Not that that seems especially likely, but it's certainly not something we ever want to be possible.

matthewwalsh0 replied:
Done, good idea.

Gudahtt commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,237 @@
+   * Includes reducer properties.
+   * For example: `{ metamask: { transactions: [] } }`.
+   */
+  getStateUI: () => unknown & { metamask: unknown };
```
This is misleading because we're not really getting the UI state here. We're getting the flattened state nested in an object under the `metamask` key.

Maybe we could build this on-the-fly using `getFlatState`, and drop this from here? Just to simplify the initial request object.

Gudahtt commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+      isHardwareWallet: isHardwareWallet(
+        uiState as Parameters<typeof isHardwareWallet>,
+      ),
+      // @ts-expect-error Mismatched types
```
Looks like it's just type errors in `SmartTransactionHook`. The missing flags are treated as optional already in practice, just not typed as such.

matthewwalsh0 replied:
Updated.

Gudahtt commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,237 @@
+ * Request to initialize and return a controller instance.
+ * Includes standard data and methods not coupled to any specific controller.
+ */
+export type ControllerInitRequest<
```
This proposal is very abstraction-heavy; we're introducing a lot of new constructs to facilitate this pattern of splitting up controller initialization. I'm wondering if we could cut any of them out and make this process simpler to understand.

If I'm understanding correctly, the current process is roughly:
* Controllers are initialized by `#initControllers` in `metamask-controller.js`
* `#initControllers` calls the `initControllers` function, passing in relevant state that only `metamask-controller.js` has (persisted state, functions, etc.)
* `initControllers` will, for each controller:
  * Create an "Init instance"
    * This instance doesn't have access to any global capabilities. It receives them when it's initialized, and uses them to complete controller-specific initialization
  * Create an "init request", and use it to initialize each "init object", which in turn initializes the controller.
    * This request passes in a custom "init messenger" and "messenger" for each controller, but otherwise is responsible for passing in the same global functions to all controllers.
  * Create a "API request", and use this to get the "controller API" functions for each controller
  * Use the return value of `getApi` and various other bits of information from the "init object" to construct a collection of all controller API methods, memstate, persisted state, and a map of controllers by name.
* Then finally, `initControllers` returns these four collections to `metamask-controller.js` to finalize initialization.

The first two steps I understand, those seem necessary (`#initControllers` for handling state, `initControllers` for performing initialization with state). But after that point I'm not sure.

I can see the benefits of having a custom "init request" per controller, in that it enables this "init messenger" pattern. We could do without an "init messenger", but it would require us to pass around controller instances to every initializer, which is not ideal for auditability (it'd make refactors of controllers much more painful). In theory we shouldn't need any special capabilities to initialize a controller, controllers can request capabilities from the messenger directly, but in practice we're a long way off from every controller behaving that way. The "init messenger" seems like a good attempt at improving the situation in the short-term, even if it does come at some cost in requiring a custom "init request" per controller.

The benefits of "init instance" and "API request" are less clear to me. Could we have an "initialize" function per controller instead, which returns the controller alongside a collection of API methods, memstate, and persisted state? What benefit does a separate "init instance" and "API request" have?

If it works just as well as a function, then this would be a good opportunity to reduce cognitive load by reducing the number of abstractions we're using here. But perhaps there is some benefit I am not seeing.

Gudahtt replied:
Ah I see, I never thought of the abstract class as a sort of template, that makes sense.

I think we could preserve the "security" benefits you referred to by passing the messengers into the init function. I've experimented with something like that here: https://github.com/MetaMask/metamask-extension/pull/29810 (I didn't update the tests yet, this is just to demonstrate the idea).

matthewwalsh0 replied:
Merged and fixed all tests and the CI, thanks!

Gudahtt replied:
Just to follow up on this, while I do see the value in the 'abstract class as a template' strategy, I am happier with the current implementation as I think the reduced complexity is a better tradeoff. The simpler init function can effectively serve as a template as well, with the complex one as further guidance. Thanks for exploring this further

Gudahtt reviewed (commented):
I like the overall structure we're using here, and a lot of the ideas for how to organize the initialization code. I left a lot of comments about the "TypeScript migration" parts of this, but my last comment on what each of these abstractions are for is probably the most relevant one for the overall plan here.

Have you considered writing an ADR about this? An ADR might be a good place to explain the overall layout/structure independently of the controller-specific details. This may help not only in explaining the pattern, but in helping others replicate it for other controllers (and especially for mobile). There are so many details in these examples, it's a bit hard to see the big picture.

Though it's great that we have such a complex example to work from, surely this hits most complexities we'll encounter in other controllers.

matthewwalsh0 committed:
Improve types

Add state types.
Improve types of getController function.
Rename getStateUI to getFlatState and deprecate.
Add detail to type errors.

matthewwalsh0 committed:
Fix env comparison

matthewwalsh0 committed:
Convert abstract class to function

matthewwalsh0 committed:
Remove state cast

matthewwalsh0 commented on app/scripts/controller-init/confirmations/ppom-controller-init.ts:
```diff
@@ -0,0 +1,74 @@
+      storageBackend: new IndexedDBPPOMStorage('PPOMDB', 1),
+      provider: getProvider(),
+      ppomProvider: {
+        // @ts-expect-error Mismatched types
```
Added detail.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/ppom-controller-init.ts:
```diff
@@ -0,0 +1,74 @@
+        PPOM: PPOMModule.PPOM,
+        ppomInit: () => PPOMModule.default(process.env.PPOM_URI),
+      },
+      state: persistedState.PPOMController as PPOMController['state'],
```
Casting removed via the use of a `Controller` union type from which we generate a `ControllerPersistedState` type.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/ppom-controller-init.ts:
```diff
@@ -0,0 +1,74 @@
+      chainId: getGlobalChainId(),
+      securityAlertsEnabled:
+        preferencesController().state.securityAlertsEnabled,
+      // @ts-expect-error Mismatched types
```
Updated.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/ppom-controller-init.ts:
```diff
@@ -0,0 +1,74 @@
+        initMessenger,
+        'PreferencesController:stateChange',
+      ),
+      cdnBaseUrl: process.env.BLOCKAID_FILE_CDN as string,
```
Updated.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/ppom-controller-init.test.ts:
The goal of these init unit tests should simply be to test all the controller initialisation logic, including any callbacks (pending migration to the messenger) passed to the controller, all API methods generated by the controller, and any client-side listeners required by the controller.

The tests added in this PR are to demonstrate how much easier it is to write these unit tests with TypeScript and better isolation, but are knowingly a subset of what should be added in order to avoid bloating the PR with unit tests.

I imagined we could add a full test suite per controller in alternate PRs, but I absolutely think the unit test aspect is a big benefit of this modularity as it allows us to verify fundamental controller and background logic that was likely never tested previously.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+
+    const controller = new TransactionController({
+      getCurrentNetworkEIP1559Compatibility: () =>
+        networkController().getEIP1559Compatibility() as Promise<boolean>,
```
Done, this was a copy and paste, but I wasn't aware we had an equivalent action in the `NetworkController`.

Also updated error comment.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+      getCurrentAccountEIP1559Compatibility: async () => true,
+      // @ts-expect-error Mismatched types
+      getExternalPendingTransactions: (address) =>
+        this.#getExternalPendingTransactions(
```
Unfortunately not, I couldn't find an equivalent action, though agree this is the logical future format.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+        networkController().getNetworkClientRegistry(...args),
+      getNetworkState: () => networkController().state,
+      // @ts-expect-error Mismatched types
+      getPermittedAccounts,
```
Added detail.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+        const globalChainId = getGlobalChainId();
+        return preferencesController().state.advancedGasFee[
+          globalChainId
+        ] as unknown as SavedGasFees;
```
Updated.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+      incomingTransactions: {
+        etherscanApiKeysByChainId: {
+          [CHAIN_IDS.MAINNET as Hex]: process.env.ETHERSCAN_API_KEY as string,
+          [CHAIN_IDS.SEPOLIA as Hex]: process.env.ETHERSCAN_API_KEY as string,
```
Removed as you're correct this is no longer required due to the accounts API migration.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+        includeTokenTransfers: false,
+        isEnabled: () =>
+          preferencesController().state.incomingTransactionsPreferences?.[
+            // @ts-expect-error Mismatched types
```
Updated.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+            )
+          ),
+      },
+      testGasFeeFlows: Boolean(process.env.TEST_GAS_FEE_FLOWS),
```
Updated, you're correct that any value would enable it, but this is for testing purposes only so there is no default in the build.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+        beforePublish: beforeTransactionPublishMMI.bind(this),
+        getAdditionalSignArguments: getAdditionalSignArgumentsMMI.bind(this),
+        ///: END:ONLY_INCLUDE_IF
+        // @ts-expect-error Mismatched types
```
Updated.

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -2556,15 +2455,13 @@ export default class MetamaskController extends EventEmitter {
       BridgeStatusController: this.bridgeStatusController,
       EnsController: this.ensController,
       ApprovalController: this.approvalController,
-      PPOMController: this.ppomController,
```
Agreed, my understanding was we can rely on a service worker that doesn't restart, though I recall some Sentry errors that suggested otherwise that may have discouraged us from removing similar restart logic in the UI initialisation.

I've done the minimum here for now as `PPOMController` is indeed persisted now so not needed here.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+
+    controllerMessenger.subscribe(
+      'TransactionController:transactionConfirmed',
+      // @ts-expect-error Mismatched types
```
Updated.

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -2444,6 +2303,46 @@ export default class MetamaskController extends EventEmitter {
+    this.ppomController = controllersByName[ControllerName.PPOMController];
+    this.txController = controllersByName[ControllerName.TransactionController];
+
+    this.controllerMessenger.subscribe(
```
Added comment.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+    ...args: Parameters<TransactionController['stopTransaction']>
+  ) {
+    await request.controller.stopTransaction(...args);
+    return request.getFlatState();
```
Done.

Moved `stopTransaction` back to the `MetamaskController` pending update of the action to not return the state.

Plus refactored `getStateUI` and `getFlatState` into a single `getFlatState` that is deprecated to discourage further use.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+    signedTransactionInHex: Hex,
+  ) {
+    const isSmartTransaction = getIsSmartTransaction(
+      uiState as Parameters<typeof getIsSmartTransaction>[0],
```
Casts removed and error comments added, plus a justification comment.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+      transactionController,
+      smartTransactionsController,
+      // @ts-expect-error Mismatched types
+      controllerMessenger,
```
This now uses the transaction controller init messenger, and is cast to the `SmartTransactionHookMessenger` as it's a superset of what is needed as we can't yet restrict further.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,396 @@
+  }
+
+  #addTransactionControllerListeners(
+    controllerMessenger: TransactionControllerInitMessenger,
```
Done, good spot.

matthewwalsh0 commented on app/scripts/controller-init/messengers/transaction-controller-messenger.ts:
```diff
@@ -0,0 +1,92 @@
+      'TransactionController:transactionNewSwapApproval',
+      'TransactionController:transactionNewSwap',
+      'TransactionController:transactionRejected',
+      'TransactionController:transactionStatusUpdated',
```
Removed.

matthewwalsh0 commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,237 @@
+ * `ControllerMessengerType` is the messenger required by the controller itself.
+ * `InitMessengerType` is the messenger required by the client to initialize the controller and register related listeners.
+ */
+export abstract class ControllerInit<
```
Class removed in favour of function, more detail in a separate PR comment.

matthewwalsh0 commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,176 @@
+  };
+}
+
+function getControllerOrThrow<T>(
```
This is now completely type safe via a `ControllerByName` type generated from a `Controller` union type in which we list all the supported controllers.

I considered making this `Controller` type a generic type as suggested in order to decouple these utilities and types from the controller types, though I haven't yet as I feared this would add further complexity due to an additional generic type passed around on many of the other types.

Plus the controller init files would already be coupled to the controller types as we now provide a correctly typed `persistedState` and `getFlatState`.

matthewwalsh0 commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,176 @@
+  const initInstances = initObjects.map((initObject) =>
+    (initObject as InitInstance).init
+      ? (initObject as InitInstance)
+      : new LegacyControllerInit(initObject as InitFunction),
```
Great point, but no longer needed as we now only accept functions.

matthewwalsh0 commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,176 @@
+/**
+ * Adapter class to handle legacy controllers that are returned from a function.
+ */
+class LegacyControllerInit extends ControllerInit<
```
Renamed in a previous commit, but removed in favour of init functions.

Gudahtt committed:
refactor: Restore messengers in controller init request

Restore the controller messenger and init messenger instances to the
controller init request. These were present in an earlier draft, but
removed in the process of converting the init class to a function.

matthewwalsh0 committed:
Fix messenger callbacks

Fix unit tests.

matthewwalsh0 committed:
Add unit tests

matthewwalsh0 commented on app/scripts/controller-init/controller-list.ts:
```diff
@@ -0,0 +1,56 @@
+import SwapsController from '../controllers/swaps';
+
+/**
+ * Union of all controllers.
```
Done.

Gudahtt commented on app/scripts/controller-init/confirmations/ppom-controller-init.test.ts:
```diff
@@ -0,0 +1,66 @@
+    const requestMock = buildInitRequestMock();
+
+    // @ts-expect-error Mocked subset of full state object
+    requestMock.getController.mockReturnValue({
```
Nit: I found this confusing to read because it wasn't clear this was meant to return a mock preferences controller specifically. Maybe this would be clearer if we use a mock implementation here instead, and verify that the controller being asked for is `PreferencesController`?

Gudahtt replied:
I see this same mock return value is in `buildInitRequestMock`. Maybe we can share the implementation

matthewwalsh0 replied:
Moved to a separate function and duplication removed.

Gudahtt commented on app/scripts/controller-init/confirmations/ppom-controller-init.test.ts:
```diff
@@ -0,0 +1,66 @@
+describe('PPOM Controller Init', () => {
+  const ppomControllerClassMock = jest.mocked(PPOMController);
+
+  function testConstructorProperty<T extends keyof PPOMControllerOptions>(
```
Nit: I struggled a bit in understanding what a "constructor property" was. I see that it's referring to a parameter of the "options" object passed to the constructor. Maybe we could make this more clear by calling this `testConstructorOption` or `testConstructorParameter`? This would also help differentiate the thing we're checking here from the `controllerProperties` passed in (at first I thought these were for the same controller).

A JSDoc comment would also be very helpful here in understanding what the tests are asserting

matthewwalsh0 replied:
Renamed and comment added.

Gudahtt commented on app/scripts/controller-init/test/utils.ts:
```diff
@@ -0,0 +1,49 @@
+    getTransactionMetricsRequest: jest.fn(),
+    initMessenger: buildControllerMessengerMock(),
+    persistedState: {},
+  } as unknown as jest.Mocked<ControllerInitRequest>;
```
Ideally we'd avoid the need to cast here, this would hide any type errors from missing properties, invalid properties, etc.

I tried getting rid of this locally, and it seems that it was only really needed because of the messengers. Even if I made this a generic function and passed through the specific messenger types, there was some conflict between the type transformation done here and in `buildControllerMessengerMock` (they conflicted despite both being mocks, maybe something to do with how that type applies to properties recursively 🤔 ).

It would be much easier if we just mocked the global functions here, and used "real" messengers instead. In controller unit tests this is usually what we do, e.g. https://github.com/MetaMask/core/blob/aec3860262a270a8cb3ed2ee668be0938243bc4b/examples/example-controllers/src/gas-prices-controller.test.ts#L140

Here we may be able to even reuse the same messenger instance we setup in the `messengers/` directory, rather than creating it anew in the test. Though either way is fine.

Messengers were designed to be used this way, as mock values. Creating a mock messenger feels a bit like creating a mock of a mock. I get the impulse to use `jest.fn()` consistently for everything, and really I don't see the harm in it either except that it seems like more work, and I don't know how to get the types to work properly.

matthewwalsh0 replied:
Switched to using real callbacks and instances.

Gudahtt commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,168 @@
+ * e.g. `{ TransactionController: { transactions: [] } }`.
+ */
+export type ControllerPersistedState = {
+  [name in ControllerName]: ControllerByName[name]['state'];
```
Not all state is persisted, so `Partial` seems appropriate here

```suggestion
  [name in ControllerName]: Partial<ControllerByName[name]['state']>;
```

In theory we could derive a more correct representation of persisted state using the controller metadata, but that's just set at runtime at the moment. Not statically in the type. Plus any old installation might have a different set of state persisted, so `Partial` is maybe a safer bet anyway.

matthewwalsh0 replied:
Done.

Gudahtt commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,168 @@
+   * Includes controller name properties.
+   * e.g. `{ TransactionController: { transactions: [] } }`.
+   */
+  persistedState: ControllerPersistedState;
```
Adding `Partial` here handles controllers that don't get persisted, and it handles the case where a user is missing persisted state for one or more controllers (because they're updating from an older version that lacked them).
```suggestion
  persistedState: Partial<ControllerPersistedState>;
```

Gudahtt replied:
Or maybe we could include the wrapper `Partial` as part of `ControllerPersistedState`? Maybe that would make more sense semantically.

Gudahtt commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,168 @@
+   * Required controller messenger instance.
+   * Generated using the callback specified in `getControllerMessengerCallback`.
+   */
+  controllerMessenger: ControllerMessengerType extends BaseRestrictedControllerMessenger
```
Not sure I follow what this conditional type was meant to achieve. The `ControllerMessengerType` type is optional, but that inevitably results in `never` here. It's both optional and not optional.

I suppose this was just for ease of testing? If so, maybe we can deal with that another way instead, and keep `ControllerMessengerType` as mandatory.

Gudahtt replied:
Though, at least for the init messenger, it does make sense for it to be optional. That wouldn't always be needed.

In that case, perhaps we can make the type less of a contradiction by only including the `initMessenger` property if the `InitMessengerType` is given.

matthewwalsh0 replied:
Done, left the init messenger as optional but switched syntax so the property no longer exists rather than having `never` type.

Gudahtt commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,304 @@
+    // @ts-expect-error Controller type does not support undefined return value
+    getPermittedAccounts,
+    // @ts-expect-error Preferences controller uses Record rather than specific type
+    getSavedGasFees: (chainId: Hex) =>
+      preferencesController().state.advancedGasFee[chainId],
```
The implementation on `main` looks more like this:

```suggestion
    getSavedGasFees: () => {
      const globalChainId = this.#getGlobalChainId();
      return preferencesController().state.advancedGasFee[globalChainId];
    },
```

Was this an intentional functional change?

matthewwalsh0 replied:
Reverted, though we likely need to update this in `main`.

Gudahtt commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -0,0 +1,304 @@
+    // @ts-expect-error Preferences controller uses Record rather than specific type
+    getSavedGasFees: (chainId: Hex) =>
+      preferencesController().state.advancedGasFee[chainId],
+    incomingTransactions: {
```
Nit: I see that `etherscanApiKeysByChainId` was removed here. Maybe best to remove that in a separate PR, so that the reason for its removal isn't lost in the git history amidst the rest of this refactor

matthewwalsh0 replied:
Reverted for clarity, though can confirm there should be zero impact.

Gudahtt commented on app/scripts/controller-init/messengers/messengers.ts:
I'm guessing you renamed this to `index.ts`, but left this by accident?

matthewwalsh0 replied:
Exactly, removed.

Gudahtt commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,168 @@
+   *
+   * @deprecated Will be removed in the future pending multi-chain support.
+   */
+  getGlobalChainId(): string;
```
Nit: Perhaps we could use the more-specific type `Hex` here? It looks like `Hex` is returned by the implementation.

```suggestion
  getGlobalChainId(): Hex;
```

matthewwalsh0 replied:
Done.

Gudahtt commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,169 @@
+  initFunctions: Partial<
+    Record<ControllersToInitialize, BaseControllerInitFunction>
+  >;
+  initRequest: Omit<ControllerInitRequest, 'getController'>;
```
Since the messenger refactor, we should be omitting two more properties:
```suggestion
  initRequest: Omit<
    ControllerInitRequest,
    'getController' | 'controllerMessenger' | 'initMessenger'
  >;
```

matthewwalsh0 replied:
Done.

Gudahtt commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,169 @@
+}: {
+  baseControllerMessenger: BaseControllerMessenger;
+  existingControllers?: Controller[];
+  initFunctions: Partial<
+    Record<ControllersToInitialize, BaseControllerInitFunction>
+  >;
```
Nit: I believe this would be a more specific/correct type:

```suggestion
  initFunctions: {
    [name in ControllersToInitialize]: ControllerInitFunction<
      ControllerByName[name],
      ReturnType<(typeof CONTROLLER_MESSENGERS)[name]['getMessenger']>,
      ReturnType<(typeof CONTROLLER_MESSENGERS)[name]['getInitMessenger']>
    >;
  };
```

Gudahtt replied:
I made this adjustment below in the `initFunctions` loop to accommodate this type change:

```typescript
  for (const [key, value] of Object.entries(initFunctions)) {
    const controllerName = key as ControllersToInitialize;
    const initFunction = value as ControllerInitFunction<
      ControllerByName[typeof controllerName],
      ReturnType<
        (typeof CONTROLLER_MESSENGERS)[typeof controllerName]['getMessenger']
      >,
      ReturnType<
        (typeof CONTROLLER_MESSENGERS)[typeof controllerName]['getInitMessenger']
      >
    >;
    ...
  }
```

Probably worth declaring this utility type on its own to avoid the repetition, e.g. `ControllerInitFunctionFromName<name extends ControllerName>`

matthewwalsh0 replied:
Done.

Gudahtt commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,169 @@
+  initFunctions,
+  initRequest,
+}: {
+  baseControllerMessenger: BaseControllerMessenger;
```
It'd be nice to use a more-correct type here as well, but I don't think we should bother attempting this until the `Controllers` type includes everything. It wouldn't be easy to make this correctly typed before then.

Gudahtt reviewed (commented):
Looks great! Left a few more suggestions on improvements, but I think this is getting pretty close to as good as it'll get.

matthewwalsh0 committed:
Update types

Use real messengers instead of mocks.

matthewwalsh0 committed:
Merge remote-tracking branch 'origin/main' into refactor/modular-controller-init

matthewwalsh0 commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,169 @@
+    // @ts-expect-error: Union too complex.
+    acc[controller.name] = controller;
+    return acc;
+  }, {} as ControllerByName);
```
Added `Partial`, renamed, stopped seeing union error after making it partial and casting the value.

matthewwalsh0 commented on app/scripts/controller-init/utils.ts:
```diff
@@ -0,0 +1,169 @@
+  ): ControllerByName[Name] => getControllerOrThrow(controllersByName, name);
+
+  for (const [name, initFunction] of Object.entries(initFunctions)) {
+    const controllerName = name as ControllersToInitialize;
```
Done.

matthewwalsh0 commented on app/scripts/controller-init/types.ts:
```diff
@@ -0,0 +1,168 @@
+  /**
+   * The key used to store the controller state in the memory-only store.
+   * Defaults to the controller `name` property if `undefined`.
+   * If `null`, the controller state will not be stored in memory.
```
Good point, done.

Gudahtt commented on app/scripts/controller-init/confirmations/ppom-controller-init.test.ts:
```diff
@@ -0,0 +1,91 @@
+
+jest.mock('@metamask/ppom-validator');
+
+function buildControllerMock(options?: Record<string, unknown>) {
+  return {
+    state: { securityAlertsEnabled: true },
+    ...options,
+  } as unknown as PPOMController;
+}
```
I see a few areas for improvement here:

* The type we're casting to here is wrong, this should be a `PreferencesController` mock.
* Nit: The purpose of this function would be much clearer with docs.
* Nit: The type of the parameter could be improved
* Nit: Our TypeScript guidelines would have us use `@ts-expect-error` here instead (this is an expected error, rather than a genuine type transformation that we know is safe to make).

Here is a suggestion including all changes:

```suggestion
/**
 * Build a mock PreferencesController.
 *
 * @param partialMock - A partial mock object for the PreferencesController, merged
 * with the default mock.
 * @returns A mock PreferencesController.
 */
function buildControllerMock(
  partialMock?: Partial<PreferencesController>,
): PreferencesController {
  const defaultPreferencesControllerMock = {
    state: { securityAlertsEnabled: true },
  };
  // @ts-expect-error Incomplete mock, just includes properties used by code-under-test.
  return {
    ...defaultPreferencesControllerMock,
    ...partialMock,
  };
}
```

Note that we need to add a new type import here for `PreferencesController` as well (the extension version, not the one from the package).

matthewwalsh0 replied:
Done, thanks!

Gudahtt commented on app/scripts/controller-init/confirmations/transaction-controller-init.test.ts:
```diff
@@ -0,0 +1,174 @@
+
+jest.mock('@metamask/transaction-controller');
+
+function buildControllerMock(options?: Record<string, unknown>) {
+  return {
+    getNetworkClientRegistry: jest.fn().mockReturnValue({}),
+    ...options,
+  } as unknown as TransactionController;
+}
```
Nit: Similar situation to the `buildControllerMock` suggestion I left in `ppom-controller-init`, the exact same problems apply here as well. Here's a suggestion to fix them:

```suggestion
/**
 * Build a mock NetworkController.
 *
 * @param partialMock - A partial mock object for the NetworkController, merged
 * with the default mock.
 * @returns A mock NetworkController.
 */
function buildControllerMock(
  partialMock?: Partial<NetworkController>,
): NetworkController {
  const defaultNetworkControllerMock = {
    getNetworkClientRegistry: jest.fn().mockReturnValue({}),
  };
  // @ts-expect-error Incomplete mock, just includes properties used by code-under-test.
  return {
    ...defaultNetworkControllerMock,
    ...partialMock,
  };
}
```

Note that we need to add a type import here as well for `NetworkController`.

matthewwalsh0 replied:
Done, thanks!

Gudahtt commented on app/scripts/controller-init/test/utils.ts:
```diff
@@ -0,0 +1,27 @@
+    getPermittedAccounts: jest.fn(),
+    getProvider: jest.fn(),
+    getTransactionMetricsRequest: jest.fn(),
+    persistedState: {} as ControllerPersistedState,
```
Nit: Unnecessary cast, `{}` is already compatible with this type since `Partial` was added
```suggestion
    persistedState: {},
```

matthewwalsh0 replied:
Done.

matthewwalsh0 committed:
Update unit test helpers

Gudahtt reviewed (approved):
LGTM! 🚀
